### PR TITLE
Support cpu limit in docker

### LINKED ISSUE #1824: support cpu limit in docker
docker can limit containers cpu usage by `docker run --cpus {num}`.
ninja 1.10 couldn't read it currently.
so if I run `docker run --cpus 2 some_image ninja`
it will run 10 sub process (my cpu is 8 core).
```
[okhowang@VM_51_247_centos /home/linuxbrew/.linuxbrew]$ docker run -it --rm --cpus 3 -v$PWD:$PWD -w$PWD ubuntu:20.04 bin/ninja --version
1.10.0
[okhowang@VM_51_247_centos /home/linuxbrew/.linuxbrew]$ docker run -it --rm --cpus 3 -v$PWD:$PWD -w$PWD ubuntu:20.04 bin/ninja --help
usage: ninja [options] [targets...]

if targets are unspecified, builds the 'default' target (see manual).

options:
  --version      print ninja version ("1.10.0")
  -v, --verbose  show all command lines while building

  -C DIR   change to DIR before doing anything else
  -f FILE  specify input build file [default=build.ninja]

  -j N     run N jobs in parallel (0 means infinity) [default=10 on this system]
  -k N     keep going until N jobs fail (0 means infinity) [default=1]
  -l N     do not start new jobs if the load average is greater than N
  -n       dry run (don't run commands but act like they succeeded)

  -d MODE  enable debugging (use '-d list' to list modes)
  -t TOOL  run a subtool (use '-t list' to list subtools)
    terminates toplevel options; further flags are passed to the tool
  -w FLAG  adjust warnings (use '-w list' to list warnings)
```
windows container has some problem when isolation is process
```
C:\ProgramData\chocolatey\lib\ninja\tools> docker run -it --rm --cpus 4 --isolation=process -v "${PWD}:$PWD" -w $PWD mcr.microsoft.com/windows/nanoserver:2004 ninja --help
usage: ninja [options] [targets...]

if targets are unspecified, builds the 'default' target (see manual).

options:
  --version      print ninja version ("1.10.0")
  -v, --verbose  show all command lines while building

  -C DIR   change to DIR before doing anything else
  -f FILE  specify input build file [default=build.ninja]

  -j N     run N jobs in parallel (0 means infinity) [default=10 on this system]
  -k N     keep going until N jobs fail (0 means infinity) [default=1]
  -l N     do not start new jobs if the load average is greater than N
  -n       dry run (don't run commands but act like they succeeded)

  -d MODE  enable debugging (use '-d list' to list modes)
  -t TOOL  run a subtool (use '-t list' to list subtools)
    terminates toplevel options; further flags are passed to the tool
  -w FLAG  adjust warnings (use '-w list' to list warnings)
C:\ProgramData\chocolatey\lib\ninja\tools> docker run -it --rm --cpus 4 --isolation=process -v "${PWD}:$PWD" -w $PWD mcr.microsoft.com/windows/nanoserver:2004 ninja --version
1.10.0
```

### PR DESCRIPTION
**Pull request by okhowang**
close #1824

on windows docker with process isolation, use job api to get cpu limit.
on linux docker, use cgroups to get cpu limit.

### CODE REVIEW

**Comment by jhasse** [comment]
Any idea why https://github.com/ninja-build/ninja/commit/1bcc689324bdee090eed035353724abc3fa7c909 did work for me?

File: src/util.cc [hunk 1]
```diff
@@ -479,6 +482,155 @@ string StripAnsiEscapeCodes(const string& in) {
...
+  int parentId;
+  StringPiece deviceId;
+  StringPiece root;
+  StringPiece mountPoint;
+  vector<StringPiece> options;
+  vector<StringPiece> optionalFields;
+  StringPiece fsType;
+  StringPiece mountSource;
+  vector<StringPiece> superOptions;
+  static bool parse(const string& line, MountPoint& self) {
```
Comment (jhasse) [inline src/util.cc:508]: Please use a normal method instead of `self`.

Comment (okhowang) [reply src/util.cc]: done

**Comment by okhowang** [comment]
> Any idea why [1bcc689](https://github.com/ninja-build/ninja/commit/1bcc689324bdee090eed035353724abc3fa7c909) did work for me?

I notice that there are many arguments about cpu-limit.
https://docs.docker.com/config/containers/resource_constraints/
`--cpus=<value>` is using cgroup to limit cpu usage.
`--cpuset-cpus` is using cpu-affinity api to limit process run on specific cpus

so [1bcc689](https://github.com/ninja-build/ninja/commit/1bcc689324bdee090eed035353724abc3fa7c909) work only when `--cpuset-cpus` is actived.
my patch work when `--cpus` is actived.

**Comment by okhowang** [comment]
> Thanks!
> 
> The detection code is quite long (especially the cgroups version). If we include this, we should definitely not run it, if the user specified a job count with `-j`.

I have modified `ninja.cc` to bypass `GuessParallelism` when `-j` set.

**Comment by jhasse** [comment]
Thanks and nice approach using the destructor to set it at the end of the scope :)

Looking forward to some further comments by others.

**Comment by rufinio** [comment]
Successfully tested `--cpus=<value>`

with following configuration:

- OS: Microsoft Windows Server 2022 Standard 10.0.20348
- Docker: version 20.10.7, build 40ef3b6
- Image: mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2022

**Comment by okhowang** [comment]
> Anyone tested on Linux, especially start up time in different scenarios? (no idea how taxing those `/proc` opens can be)

I tested with `time ninja --version`(--version will read cpu limit config) on my WSL2 with version 1.10.0 and this MR.

my version
```
$ time ./ninja --version
1.10.2.git

real    0m0.023s
user    0m0.000s
sys     0m0.004s
```
system ninja
```
time ninja --version
1.10.0

real    0m0.002s
user    0m0.000s
sys     0m0.001s
```

**Comment by rufinio** [comment]
@jhasse: Anything else to get this PR merged?

**Pull request merged**